tmt/utils/jira.py

```diff
@@ -0,0 +1,134 @@
+from tmt.plugins import ModuleImporter
+
+if TYPE_CHECKING:
+    import jira
+
+
+import_jira: ModuleImporter['jira'] = ModuleImporter(  # type: ignore[valid-type]
+    'jira',
+    tmt.utils.ReportError,
+    "Install 'tmt+link-jira' to use the Jira linking.")
+
+
+def jira_link(
+        nodes: Sequence[Union['tmt.base.Test', 'tmt.base.Plan', 'tmt.base.Story']],
+        links: 'tmt.base.Links',
+        logger: tmt.log.Logger,
+        separate: bool = False) -> None:
+    """ Link the object to Jira issue and create the URL to tmt web service """
+    jira_module = import_jira(logger)
+
+    def create_url_params(tmt_object: tmt.base.Core) -> dict[str, Any]:
+        """ Prepare url parameters prefixed with tmt object type """
+        tmt_type = tmt_object.__class__.__name__.lower()
+        fmf_id = tmt_object.fmf_id
+
+        url_params: dict[str, Any] = {
+            'format': 'html',
+            f'{tmt_type}-url': fmf_id.url,
+            f'{tmt_type}-name': fmf_id.name
+            }
+
+        if fmf_id.path and isinstance(fmf_id.git_root, Path):
+            fmf_path = fmf_id.path.relative_to(fmf_id.git_root).resolve().as_posix()
+            fmf_path = fmf_path.removeprefix(fmf_id.git_root.as_posix())
+            url_params[f'{tmt_type}-path'] = fmf_path.lstrip('/')
+
+        if fmf_id.ref:
+            url_params[f'{tmt_type}-ref'] = fmf_id.ref
+
+        return url_params
+
+    def create_url(baseurl: str, url_params: dict[str, str]) -> str:
+        """ Prepare the full url from base and parameters """
+        return urllib.parse.urljoin(baseurl, '?' + urllib.parse.urlencode(url_params))
+
+    # Setup config tree, exit if config is missing
+    try:
+        config_tree = tmt.utils.Config()
+        linking_node = cast(Optional[fmf.Tree], config_tree.fmf_tree.find('/user/linking'))
+    except tmt.utils.MetadataError:
+        return
+
+    if not linking_node:
+        # Linking is not setup in config, therefore user does not want to use linking
+        return
+
+    logger.print(
+        f'Linking {fmf.utils.listed([type(node).__name__.lower() for node in nodes])}'
+        f' to Jira issue.')
+
+    issue_trackers: Any = linking_node.data.get('issue-tracker')
+
+    if not issue_trackers:
+        # Incorrectly setup config - wrong keyword
+        raise tmt.utils.GeneralError("Invalid config!")
+
+    if not isinstance(issue_trackers[0], dict):
+        # Incorrectly setup config - configuration empty
+        raise tmt.utils.GeneralError("Invalid config!")
+
+    linking_config = cast(list[dict[str, Any]], issue_trackers)[0]
+
+    verifies = links.get('verifies')[0]
+    target = verifies.to_dict()['target']
+    # Parse the target url
+    issue_id = target.split('/')[-1]
+    # ignore[attr-defined]: it is defined, but mypy seems to fail
+    # detecting it correctly.
+    jira = jira_module.JIRA(  # type: ignore[attr-defined]
+        server=linking_config['url'],
+        token_auth=linking_config['token'])
```

Comment:
Looks like github didn't save my comment, so here we go again...  
How are we verifying that url, token is present and is string? Is this a job for a schema or can we create a typeddict for example?
```
class JiraConfig(TypedDict):
    url: str
    token: str
    tmt_web_url: str
```
and use it here:
```
    try:
        linking_config = JiraConfig(
            url=jira_config['url'],
            token=jira_config['token'],
            tmt_web_url=jira_config['tmt-web-url']
        )
    except KeyError as e:
        raise tmt.utils.GeneralError(f"Invalid Jira config: Missing required key {str(e)}")
```
Or am I missing something? I presume the potential base IssueTrackerConfig class would be elsewhere, but where? 
Not blocking comment of course ;)

Reply:
Yes, the proper & nice checking should go to the config implementation together with a schema. Let's not block on that. For now, I'd say we should do at least a basic check for the correct types.

Reply:
This has been covered in 7ecd3bb7.